fendermint/vm/interpreter/src/signed.rs

```diff
@@ -17,6 +20,67 @@ pub struct InvalidSignature(pub String);
 pub type SignedMessageApplyRet = Result<FvmApplyRet, InvalidSignature>;
 pub type SignedMessageCheckRet = Result<FvmCheckRet, InvalidSignature>;
 
+/// Different kinds of signed messages.
+///
+/// This technical construct was introduced so we can have a simple linear interpreter stack
+/// where everything flows through all layers, which means to pass something to the FVM we
+/// have to go through the signature check.
+pub enum VerifiableMessage {
+    /// A normal message sent by a user.
+    Signed(SignedMessage),
+    /// Something we constructed to pass on to the FVM.
+    Synthetic(SyntheticMessage),
+}
+
+impl VerifiableMessage {
+    pub fn verify(&self, chain_id: &ChainID) -> Result<(), SignedMessageError> {
+        match self {
+            Self::Signed(m) => m.verify(chain_id),
+            Self::Synthetic(m) => m.verify(chain_id),
+        }
+    }
+
+    pub fn into_message(self) -> FvmMessage {
+        match self {
+            Self::Signed(m) => m.into_message(),
+            Self::Synthetic(m) => m.message,
+        }
+    }
+}
+
+pub struct SyntheticMessage {
+    /// The artifical message.
+    message: FvmMessage,
+    /// The CID of the original message (assuming here that that's what was signed).
+    orig_cid: cid::Cid,
+    /// The signature over the original CID.
+    signature: Signature,
```

Comment:
Just to double-check, this is the original signature of the message, right? Should we maybe call it `WrappedMessage` instead of `SyntheticMessage`? (weak opinion weakly held, but synthetic seemed like "artificial" to me :) )

Reply:
It _is_ an artificial message, that's exactly what I was trying to convey, that it's just a transient technical message to piggy-back on the signature checks and FVM message execution. The construction of the FVM message from the original is synthesising an artificial message that nobody sent.

It's indeed wrapping something but that wouldn't say why; so does the other one after all. 

You are correct that `signature` is the signature over the original message, in particular the CID of the original message. 

We could get rid of this if we used FVM messages, in which case instead of synthesising a message in the `ChainMessageInterpreter` from an IPC specific one, we would parse an FVM message into an IPC message to look for CIDs to resolve.

Reply:
I think this is fine for now, and now I understand the rationale behind the name. Thanks!